compiler/ilgen/VirtualMachineOperandStack.hpp

```diff
@@ -0,0 +1,190 @@
+#include "ilgen/VirtualMachineRegister.hpp"
+#include "ilgen/IlBuilder.hpp"
+
+namespace TR { class MethodBuilder; }
+
+namespace OMR
+{
+
+/**
+ * @brief simulates an operand stack used by many bytecode based virtual machines
+ * In such virtual machines, the operand stack holds the intermediate expression values
+ * computed by the bytecodes. Ther compiler simulates this operand stack as well, but
+ * what is pushed to and popped from the simulated operand stack are expression nodes
+ * that represent the value computed by the bytecodes. As each bytecode pops expression
+ * nodes off the operand stack, it combines them to form more complicated expressions
+ * which are then pushed back onto the operand stack for consumption by other bytecodes.
+ *
+ * The stack is represented as an array of pointers to TR::IlValue's, making it
+ * easy to use IlBuilder services to consume and compute new values. Note that,
+ * unlike VirtualMachineRegister, the simulated operand stack is *not* maintained
+ * by the method code as part of the method's stack frame. This approach requires
+ * modelling the state of the operand stack at all program points, which means
+ * there cannot only be one VirtualMachineOperandStack object.
+ *
+ * The current implementation does not share anything among different
+ * VirtualMachineOperandStack objects. Possibly, some of the state could be
+ * shared to save some memory. For now, simplicity is the goal.
+ *
+ * VirtualMachineOperandStack implements VirtualMachineState:
+ * Commit() simply iterates over the simulated operand stack and stores each
+ *   value onto the virtual machine's operand stack (more details at definition).
+ * Reload() is left empty; assumption is that each BytecodeBuilder handler will
+ *   update the state of the operand stack appropriately on return from the
+ *   interpreter.
+ * MakeCopy() copies the state of the operand stack
+ * MergeInto() is slightly subtle. Operations may have been already created
+ *   below the merge point, and those operations will have assumed the
+ *   expressions are stored in the TR::IlValue's for the state being Merged
+ *   *to*. So the purpose of Merge() is to store the values of the current
+ *   state into the same variables as in the "other" state.
+ * 
+ * VirtualMachineOperandStack provides several stack-y operations:
+ *   Push() pushes a TR::IlValue onto the stack
+ *   Pop() pops and returns a TR::IlValue from the stack
+ *   Top() returns the TR::IlValue at the top of the stack
+ *   Pick() returns the TR::IlValue "depth" elements from the top
+ *   Drop() discards "depth" elements from the stack
+ *   Dup() is a convenience function for Push(Top())
+ *
+ * To facilitate propagating an operand stack as control flows from one bytecode
+ * to another, VirtualMachineOperandStack provides a function MakeCopy() which
+ * allocates a new VirtualMachineOperandStack and makes it an identical
+ * copy of the current operand stack. It is typically used to 
```

Comment:
actually, that whole paragraph is an earlier (obviously incomplete) version of the comment. i've removed it.